framework/include/base/FEProblem.h

```diff
@@ -481,12 +481,12 @@ class FEProblem :
    * @return Const reference to the user object
    */
   template <class T>
-  const T & getUserObject(const std::string & name)
+  const T & getUserObject(const std::string & name, unsigned int tid = 0)
```

Comment:
No. It should work even if those are different threaded loops. The identity of the thread is meaningless. What matters is the _tid, which is used to look up the prepared elements etc. If those are the same in both objects they will access consistent data!